src/Symfony/Bundle/FrameworkBundle/CacheWarmer/ControllerMetadataCacheWarmer.php

```diff
@@ -0,0 +1,137 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\CacheWarmer;
+
+use Symfony\Component\Config\ConfigCacheFactoryInterface;
+use Symfony\Component\Config\ConfigCacheInterface;
+use Symfony\Component\Config\Resource\FileResource;
+use Symfony\Component\Filesystem\Filesystem;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpKernel\CacheWarmer\CacheWarmerInterface;
+use Symfony\Component\HttpKernel\Controller\ControllerResolverInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ControllerMetadataFactoryInterface;
+use Symfony\Component\HttpKernel\ControllerMetadata\ControllerMetadataUtil;
+use Symfony\Component\Routing\Route;
+use Symfony\Component\Routing\RouterInterface;
+
+/**
+ * Generates the cache for controller metadata.
+ *
+ * @author Iltar van der Berg <kjarli@gmail.com>
+ */
+final class ControllerMetadataCacheWarmer implements CacheWarmerInterface
+{
+    private $router;
+    private $controllerResolver;
+    private $controllerMetadataFactory;
```

Comment:
Your IDE can already determine the type via constructor injection. The most recent component doesn't add type-hints either: https://github.com/symfony/symfony/tree/master/src/Symfony/Component/Cache